php84Packages.composer: 2.8.1 -> 2.8.4

drupol opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

drupol committed:
php84Packages.composer: 2.8.1 -> 2.8.4

This release includes an extra patch to fix reproducible outputs, see https://github.com/composer/composer/pull/12090

drupol committed:
php84Packages.psysh: 0.12.4 -> 0.12.7

drupol committed:
phpdocumentor: minor cleanup

piotrkwiecinski commented:
Does the composer load order fix affect hashes of any packages using builder2?

drupol commented:
> Does the composer load order fix affect hashes of any packages using builder2?

I just tried a few things, and nothing has changed (PHPUnit, PHPStan, PSalm).

piotrkwiecinski commented on pkgs/development/php-packages/composer/default.nix:
```diff
@@ -86,7 +97,7 @@ stdenvNoCC.mkDerivation (finalAttrs: {
     outputHashMode = "recursive";
     outputHashAlgo = "sha256";
-    outputHash = "sha256-FfFwx5E2LVDSqo2P31fqtvk2P30XnTm+TUqhNSHTt/M=";
+    outputHash = "sha256-RuHBpfU/QDIsVG5WvqfZqEiXStdtATyjyKmModR4PJQ=";
```
When building this locally I get:
```
error: hash mismatch in fixed-output derivation '/nix/store/v80nb8gymjpn5k9li4ssza4b3qihnkmr-composer-vendor-2.8.4.drv':
         specified: sha256-RuHBpfU/QDIsVG5WvqfZqEiXStdtATyjyKmModR4PJQ=
            got:    sha256-McyO3Z4PSyC6LiWt8rsXziAIbEqOhiaT77gUdzZ6tzw=
```

drupol replied:
You were right, I missed that one somehow.

drupol committed:
phpPackages.psalm: 5.25.0 -> 5.26.1

drupol commented:
There's indeed a hash change for all the PHP derivations since the file generated by the autoloaders has been updated. Here's the diff between a build of PHPUnit with Composer 2.8.1 and Composer 2.8.4(+patch):

```
❯ nix shell nixpkgs#diffoscopeMinimal
❯ diffoscope result-phpunit-composer-2-8-1 result-phpunit-composer-2-8-4
--- result-phpunit-composer-2-8-1
+++ result-phpunit-composer-2-8-4
├── stat {}
│ @@ -1,7 +1,7 @@
│  
│    Size: 58           Blocks: 0          IO Block: 4096   symbolic link
│  Device: 259,4        Access: (0777/lrwxrwxrwx)  Uid: ( 1000/     pol)   Gid: (  100/   users)
│  
│ -Modify: 2025-01-10 17:47:49.888938828 +0000
│ +Modify: 2025-01-10 17:48:41.563959336 +0000
│   --- result-phpunit-composer-2-8-1/share
├── +++ result-phpunit-composer-2-8-4/share
│ │   --- result-phpunit-composer-2-8-1/share/php
│ ├── +++ result-phpunit-composer-2-8-4/share/php
│ │ │   --- result-phpunit-composer-2-8-1/share/php/phpunit
│ │ ├── +++ result-phpunit-composer-2-8-4/share/php/phpunit
│ │ │ │   --- result-phpunit-composer-2-8-1/share/php/phpunit/vendor
│ │ │ ├── +++ result-phpunit-composer-2-8-4/share/php/phpunit/vendor
│ │ │ │ │   --- result-phpunit-composer-2-8-1/share/php/phpunit/vendor/composer
│ │ │ │ ├── +++ result-phpunit-composer-2-8-4/share/php/phpunit/vendor/composer
│ │ │ │ │ │   --- result-phpunit-composer-2-8-1/share/php/phpunit/vendor/composer/InstalledVersions.php
│ │ │ │ │ ├── +++ result-phpunit-composer-2-8-4/share/php/phpunit/vendor/composer/InstalledVersions.php
│ │ │ │ │ │ @@ -318,25 +318,28 @@
│ │ │ │ │ │      private static function getInstalled()
│ │ │ │ │ │      {
│ │ │ │ │ │          if (null === self::$canGetVendors) {
│ │ │ │ │ │              self::$canGetVendors = method_exists('Composer\Autoload\ClassLoader', 'getRegisteredLoaders');
│ │ │ │ │ │          }
│ │ │ │ │ │  
│ │ │ │ │ │          $installed = array();
│ │ │ │ │ │ +        $copiedLocalDir = false;
│ │ │ │ │ │  
│ │ │ │ │ │          if (self::$canGetVendors) {
│ │ │ │ │ │              foreach (ClassLoader::getRegisteredLoaders() as $vendorDir => $loader) {
│ │ │ │ │ │                  if (isset(self::$installedByVendor[$vendorDir])) {
│ │ │ │ │ │                      $installed[] = self::$installedByVendor[$vendorDir];
│ │ │ │ │ │                  } elseif (is_file($vendorDir.'/composer/installed.php')) {
│ │ │ │ │ │                      /** @var array{root: array{name: string, pretty_version: string, version: string, reference: string|null, type: string, install_path: string, aliases: string[], dev: bool}, versions: array<string, array{pretty_version?: string, version?: string, reference?: string|null, type?: string, install_path?: string, aliases?: string[], dev_requirement: bool, replaced?: string[], provided?: string[]}>} $required */
│ │ │ │ │ │                      $required = require $vendorDir.'/composer/installed.php';
│ │ │ │ │ │ +                    self::$installedByVendor[$vendorDir] = $required;
│ │ │ │ │ │ +                    $installed[] = $required;
│ │ │ │ │ │ +                    if (strtr($vendorDir.'/composer', '\\', '/') === strtr(__DIR__, '\\', '/')) {
│ │ │ │ │ │ +                        self::$installed = $required;
│ │ │ │ │ │ +                        $copiedLocalDir = true;
│ │ │ │ │ │ -                    $installed[] = self::$installedByVendor[$vendorDir] = $required;
│ │ │ │ │ │ -                    if (null === self::$installed && strtr($vendorDir.'/composer', '\\', '/') === strtr(__DIR__, '\\', '/')) {
│ │ │ │ │ │ -                        self::$installed = $installed[count($installed) - 1];
│ │ │ │ │ │                      }
│ │ │ │ │ │                  }
│ │ │ │ │ │              }
│ │ │ │ │ │          }
│ │ │ │ │ │  
│ │ │ │ │ │          if (null === self::$installed) {
│ │ │ │ │ │              // only require the installed.php file if this file is loaded from its dumped location,
│ │ │ │ │ │ @@ -346,14 +349,14 @@
│ │ │ │ │ │                  $required = require __DIR__ . '/installed.php';
│ │ │ │ │ │                  self::$installed = $required;
│ │ │ │ │ │              } else {
│ │ │ │ │ │                  self::$installed = array();
│ │ │ │ │ │              }
│ │ │ │ │ │          }
│ │ │ │ │ │  
│ │ │ │ │ │ +        if (self::$installed !== array() && !$copiedLocalDir) {
│ │ │ │ │ │ -        if (self::$installed !== array()) {
│ │ │ │ │ │              $installed[] = self::$installed;
│ │ │ │ │ │          }
│ │ │ │ │ │  
│ │ │ │ │ │          return $installed;
│ │ │ │ │ │      }
│ │ │ │ │ │  }
│ │ │ │ │ │ ├── stat {}
│ │ │ │ │ │ │ @@ -1,7 +1,7 @@
│ │ │ │ │ │ │  
│ │ │ │ │ │ │ +  Size: 16222            Blocks: 32         IO Block: 4096   regular file
│ │ │ │ │ │ │ -  Size: 16143            Blocks: 32         IO Block: 4096   regular file
│ │ │ │ │ │ │  Device: 259,1    Access: (0444/-r--r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
│ │ │ │ │ │ │  
│ │ │ │ │ │ │  Modify: 1970-01-01 00:00:01.000000000 +0000
❯
```

drupol committed:
phpunit: update hash

drupol committed:
phpPackages.phpstan: update hash

drupol committed:
phpPackages.php-cs-fixer: 3.64.0 -> 3.67.0

drupol committed:
build-support/php: improve PHP builder

This update provide a new feature that allows to build packages without `composer.lock` file (e.g. PHP CodeSniffer).

drupol committed:
phpPackages.php-codesniffer: 3.11.0 -> 3.11.2

piotrkwiecinski commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 372696`

---
### `x86_64-linux`
<details>
  <summary>:x: 4 packages failed to build:</summary>
  <ul>
    <li>php81Packages.psysh</li>
    <li>php82Packages.psysh</li>
    <li>php83Packages.psysh</li>
    <li>php84Packages.psysh</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 94 packages built:</summary>
  <ul>
    <li>adminer</li>
    <li>agorakit</li>
    <li>bookstack</li>
    <li>composer-require-checker</li>
    <li>davis</li>
    <li>firefly-iii</li>
    <li>firefly-iii-data-importer</li>
    <li>flarum</li>
    <li>grocy</li>
    <li>invoiceplane</li>
    <li>kimai</li>
    <li>laravel</li>
    <li>librenms</li>
    <li>movim</li>
    <li>n98-magerun</li>
    <li>n98-magerun2</li>
    <li>paratest</li>
    <li>pdepend</li>
    <li>pest</li>
    <li>phel</li>
    <li>php81Packages.box (php82Packages.box ,php83Packages.box ,php84Packages.box)</li>
    <li>php81Packages.castor</li>
    <li>php81Packages.composer</li>
    <li>php81Packages.composer-local-repo-plugin</li>
    <li>php81Packages.cyclonedx-php-composer</li>
    <li>php81Packages.deployer</li>
    <li>php81Packages.grumphp</li>
    <li>php81Packages.phan</li>
    <li>php81Packages.phing</li>
    <li>php81Packages.phive</li>
    <li>php81Packages.php-codesniffer</li>
    <li>php81Packages.php-cs-fixer</li>
    <li>php81Packages.php-parallel-lint</li>
    <li>php81Packages.phpinsights</li>
    <li>php81Packages.phpmd</li>
    <li>php81Packages.phpstan</li>
    <li>php81Packages.psalm</li>
    <li>php82Packages.castor</li>
    <li>php82Packages.composer</li>
    <li>php82Packages.composer-local-repo-plugin</li>
    <li>php82Packages.cyclonedx-php-composer</li>
    <li>php82Packages.deployer</li>
    <li>php82Packages.grumphp</li>
    <li>php82Packages.phan</li>
    <li>php82Packages.phing</li>
    <li>php82Packages.phive</li>
    <li>php82Packages.php-codesniffer</li>
    <li>php82Packages.php-cs-fixer</li>
    <li>php82Packages.php-parallel-lint</li>
    <li>php82Packages.phpinsights</li>
    <li>php82Packages.phpmd</li>
    <li>php82Packages.phpstan</li>
    <li>php82Packages.psalm</li>
    <li>php83Packages.castor</li>
    <li>php83Packages.composer</li>
    <li>php83Packages.composer-local-repo-plugin</li>
    <li>php83Packages.cyclonedx-php-composer</li>
    <li>php83Packages.deployer</li>
    <li>php83Packages.grumphp</li>
    <li>php83Packages.phan</li>
    <li>php83Packages.phing</li>
    <li>php83Packages.phive</li>
    <li>php83Packages.php-codesniffer</li>
    <li>php83Packages.php-cs-fixer</li>
    <li>php83Packages.php-parallel-lint</li>
    <li>php83Packages.phpinsights</li>
    <li>php83Packages.phpmd</li>
    <li>php83Packages.phpstan</li>
    <li>php83Packages.psalm</li>
    <li>php84Packages.castor</li>
    <li>php84Packages.composer</li>
    <li>php84Packages.composer-local-repo-plugin</li>
    <li>php84Packages.cyclonedx-php-composer</li>
    <li>php84Packages.deployer</li>
    <li>php84Packages.grumphp</li>
    <li>php84Packages.phan</li>
    <li>php84Packages.phing</li>
    <li>php84Packages.phive</li>
    <li>php84Packages.php-codesniffer</li>
    <li>php84Packages.php-cs-fixer</li>
    <li>php84Packages.php-parallel-lint</li>
    <li>php84Packages.phpinsights</li>
    <li>php84Packages.phpmd</li>
    <li>php84Packages.phpstan</li>
    <li>php84Packages.psalm</li>
    <li>phpactor</li>
    <li>phpdocumentor</li>
    <li>phpunit</li>
    <li>pixelfed</li>
    <li>roave-backward-compatibility-check</li>
    <li>robo</li>
    <li>simplesamlphp</li>
    <li>snipe-it</li>
    <li>vimPlugins.phpactor</li>
  </ul>
</details>

drupol commented:
@piotrkwiecinski : I made some changes to the builder, car you try again to make sure it doesn't break anything ?

piotrkwiecinski commented:
Result from the latest version rest builds fine:
```
error: hash mismatch in fixed-output derivation '/nix/store/9gb0vf8y2y260f4bghxkvzxdsb8ds9gk-psysh-composer-repository-0.12.7.drv':
         specified: sha256-i9x/4dJ798iwMciUdO6KdCkna3SrH+L1bKEpen4Ol00=
            got:    sha256-ODUfR7PsM1YKkEIl4KEAHcY2irqlqMGlpvmEYV1M2jk=
```

drupol commented:
> Result from the latest version rest builds fine:
> ```
> error: hash mismatch in fixed-output derivation '/nix/store/9gb0vf8y2y260f4bghxkvzxdsb8ds9gk-psysh-composer-repository-0.12.7.drv':
>          specified: sha256-i9x/4dJ798iwMciUdO6KdCkna3SrH+L1bKEpen4Ol00=
>             got:    sha256-ODUfR7PsM1YKkEIl4KEAHcY2irqlqMGlpvmEYV1M2jk=
> ```
> It's from the commit `c762796`.

For which derivation ?

piotrkwiecinski commented:
It's the hash issue for php81Packages.psysh, php82Packages.psysh, php83Packages.psysh, php84Packages.psysh

```
checking for references to /build/ in /nix/store/vkvh16i2657w49a4farfp7dws27rwdc4-psysh-composer-repository-0.12.7...
error: hash mismatch in fixed-output derivation '/nix/store/9gb0vf8y2y260f4bghxkvzxdsb8ds9gk-psysh-composer-repository-0.12.7.drv':
         specified: sha256-i9x/4dJ798iwMciUdO6KdCkna3SrH+L1bKEpen4Ol00=
            got:    sha256-ODUfR7PsM1YKkEIl4KEAHcY2irqlqMGlpvmEYV1M2jk=
error: 1 dependencies of derivation '/nix/store/b6q8yp7wmyvi6ql9x9dvmw0hjm7896qd-psysh-0.12.7.drv' failed to build

shrinking RPATHs of ELF executables and libraries in /nix/store/vkvh16i2657w49a4farfp7dws27rwdc4-psysh-composer-repository-0.12.7
checking for references to /build/ in /nix/store/vkvh16i2657w49a4farfp7dws27rwdc4-psysh-composer-repository-0.12.7...
error: hash mismatch in fixed-output derivation '/nix/store/yqg2kr66jmranwnrqcca7c0msjx3qmx8-psysh-composer-repository-0.12.7.drv':
         specified: sha256-i9x/4dJ798iwMciUdO6KdCkna3SrH+L1bKEpen4Ol00=
            got:    sha256-ODUfR7PsM1YKkEIl4KEAHcY2irqlqMGlpvmEYV1M2jk=
error: 1 dependencies of derivation '/nix/store/hxlh9906z9ligs8jsi0hb7kdw81i8a2x-psysh-0.12.7.drv' failed to build

```

drupol commented:
@piotrkwiecinski Feel free to run again `nipkgs-review` now :)

piotrkwiecinski commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 372696`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 98 packages built:</summary>
  <ul>
    <li>adminer</li>
    <li>agorakit</li>
    <li>bookstack</li>
    <li>composer-require-checker</li>
    <li>davis</li>
    <li>firefly-iii</li>
    <li>firefly-iii-data-importer</li>
    <li>flarum</li>
    <li>grocy</li>
    <li>invoiceplane</li>
    <li>kimai</li>
    <li>laravel</li>
    <li>librenms</li>
    <li>movim</li>
    <li>n98-magerun</li>
    <li>n98-magerun2</li>
    <li>paratest</li>
    <li>pdepend</li>
    <li>pest</li>
    <li>phel</li>
    <li>php81Packages.box (php82Packages.box ,php83Packages.box ,php84Packages.box)</li>
    <li>php81Packages.castor</li>
    <li>php81Packages.composer</li>
    <li>php81Packages.composer-local-repo-plugin</li>
    <li>php81Packages.cyclonedx-php-composer</li>
    <li>php81Packages.deployer</li>
    <li>php81Packages.grumphp</li>
    <li>php81Packages.phan</li>
    <li>php81Packages.phing</li>
    <li>php81Packages.phive</li>
    <li>php81Packages.php-codesniffer</li>
    <li>php81Packages.php-cs-fixer</li>
    <li>php81Packages.php-parallel-lint</li>
    <li>php81Packages.phpinsights</li>
    <li>php81Packages.phpmd</li>
    <li>php81Packages.phpstan</li>
    <li>php81Packages.psalm</li>
    <li>php81Packages.psysh</li>
    <li>php82Packages.castor</li>
    <li>php82Packages.composer</li>
    <li>php82Packages.composer-local-repo-plugin</li>
    <li>php82Packages.cyclonedx-php-composer</li>
    <li>php82Packages.deployer</li>
    <li>php82Packages.grumphp</li>
    <li>php82Packages.phan</li>
    <li>php82Packages.phing</li>
    <li>php82Packages.phive</li>
    <li>php82Packages.php-codesniffer</li>
    <li>php82Packages.php-cs-fixer</li>
    <li>php82Packages.php-parallel-lint</li>
    <li>php82Packages.phpinsights</li>
    <li>php82Packages.phpmd</li>
    <li>php82Packages.phpstan</li>
    <li>php82Packages.psalm</li>
    <li>php82Packages.psysh</li>
    <li>php83Packages.castor</li>
    <li>php83Packages.composer</li>
    <li>php83Packages.composer-local-repo-plugin</li>
    <li>php83Packages.cyclonedx-php-composer</li>
    <li>php83Packages.deployer</li>
    <li>php83Packages.grumphp</li>
    <li>php83Packages.phan</li>
    <li>php83Packages.phing</li>
    <li>php83Packages.phive</li>
    <li>php83Packages.php-codesniffer</li>
    <li>php83Packages.php-cs-fixer</li>
    <li>php83Packages.php-parallel-lint</li>
    <li>php83Packages.phpinsights</li>
    <li>php83Packages.phpmd</li>
    <li>php83Packages.phpstan</li>
    <li>php83Packages.psalm</li>
    <li>php83Packages.psysh</li>
    <li>php84Packages.castor</li>
    <li>php84Packages.composer</li>
    <li>php84Packages.composer-local-repo-plugin</li>
    <li>php84Packages.cyclonedx-php-composer</li>
    <li>php84Packages.deployer</li>
    <li>php84Packages.grumphp</li>
    <li>php84Packages.phan</li>
    <li>php84Packages.phing</li>
    <li>php84Packages.phive</li>
    <li>php84Packages.php-codesniffer</li>
    <li>php84Packages.php-cs-fixer</li>
    <li>php84Packages.php-parallel-lint</li>
    <li>php84Packages.phpinsights</li>
    <li>php84Packages.phpmd</li>
    <li>php84Packages.phpstan</li>
    <li>php84Packages.psalm</li>
    <li>php84Packages.psysh</li>
    <li>phpactor</li>
    <li>phpdocumentor</li>
    <li>phpunit</li>
    <li>pixelfed</li>
    <li>roave-backward-compatibility-check</li>
    <li>robo</li>
    <li>simplesamlphp</li>
    <li>snipe-it</li>
    <li>vimPlugins.phpactor</li>
  </ul>
</details>

drupol commented:
Let's get this in!
